src/Compilers/CSharp/Portable/Symbols/FileIdentifier.cs

```diff
@@ -9,33 +9,91 @@
-    public string DisplayFilePath { get; init; }
+    private string? _filePath;
 
-    public static FileIdentifier Create(SyntaxTree tree)
-        => Create(tree.FilePath);
+    private string? _encoderFallbackErrorMessage;
+    private string? _displayFilePath;
+    private ImmutableArray<byte> _filePathChecksumOpt;
 
-    public static FileIdentifier Create(string filePath)
+    private FileIdentifier(string filePath)
+    {
+        _filePath = filePath;
+    }
+
+    private FileIdentifier(ImmutableArray<byte> filePathChecksumOpt, string displayFilePath)
+    {
+        _filePathChecksumOpt = filePathChecksumOpt;
+        _displayFilePath = displayFilePath;
+    }
+
+    private void EnsureInitialized()
     {
-        string? encoderFallbackErrorMessage = null;
-        ImmutableArray<byte> hash = default;
-        try
+        var filePath = _filePath;
+        if (filePath is not null)
         {
-            var encodedFilePath = s_encoding.GetBytes(filePath);
-            using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
-            hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            string? encoderFallbackErrorMessage = null;
+            ImmutableArray<byte> hash = default;
+            try
+            {
+                var encodedFilePath = s_encoding.GetBytes(filePath);
+                using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
+                hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            }
+            catch (EncoderFallbackException ex)
+            {
+                encoderFallbackErrorMessage = ex.Message;
+            }
+
+            var displayFilePath = GeneratedNames.GetDisplayFilePath(filePath);
+
+            _encoderFallbackErrorMessage = encoderFallbackErrorMessage;
+            _displayFilePath = displayFilePath;
+
+            ImmutableInterlocked.InterlockedInitialize(ref _filePathChecksumOpt, hash);
+            _filePath = null;
         }
-        catch (EncoderFallbackException ex)
+    }
+
+    public string DisplayFilePath
+    {
+        get
         {
-            encoderFallbackErrorMessage = ex.Message;
+            EnsureInitialized();
+
+            return _displayFilePath!;
         }
+    }
 
-        var displayFilePath = GeneratedNames.GetDisplayFilePath(filePath);
-        return new FileIdentifier { EncoderFallbackErrorMessage = encoderFallbackErrorMessage, FilePathChecksumOpt = hash, DisplayFilePath = displayFilePath };
+    public string? EncoderFallbackErrorMessage
+    {
+        get
+        {
+            EnsureInitialized();
+
+            return _encoderFallbackErrorMessage;
+        }
+    }
+
+    public ImmutableArray<byte> FilePathChecksumOpt
+    {
+        get
+        {
+            EnsureInitialized();
+
+            return _filePathChecksumOpt;
+        }
     }
+
+    public static FileIdentifier Create(SyntaxTree tree)
```

Comment:
>Create [](http://example.com/codeflow?start=33&length=6)

Is this helper used?

Reply:
![image](https://github.com/dotnet/roslyn/assets/6785178/d8f942aa-4d7d-47d2-b770-4b545da69e55)